File: frontend/src/schemas/run_progress.ts
import { z } from "zod";
import { runStatusSchema } from "./invoice_read";
import { jsonValueSchema } from "./ledger";

export const invoiceNodeSchema = z.enum([
  "Ingest", "Extract", "Validate", "Match3Way", "Policy", "Gate",
  "AutoApprove", "ExceptionTriage", "HumanReview", "Archive", "Reject",
]);
const timestamp = z.string().datetime({ offset: true });

export const nodeProgressSchema = z.object({
  name: invoiceNodeSchema,
  observed_at: timestamp.nullable(),
  event_type: z.string().nullable(),
  state: z.record(z.string(), jsonValueSchema),
}).strict();

export const runProgressSchema = z.object({
  run_id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  status: runStatusSchema,
  graph_version: z.string(),
  active_node: invoiceNodeSchema.nullable(),
  progress_source: z.literal("audit-ledger"),
  nodes: z.array(nodeProgressSchema).length(11),
  started_at: timestamp.nullable(),
  completed_at: timestamp.nullable(),
  read_at: timestamp,
}).strict();

export type RunProgress = z.infer<typeof runProgressSchema>;
export type NodeProgress = z.infer<typeof nodeProgressSchema>;

File: src/invoiceops_agent/api/schemas/run_progress.py
"""A bounded, audited view of invoice workflow progress."""

from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, JsonValue

from invoiceops_agent.api.schemas.invoice_read import RunStatus
from invoiceops_agent.graph.state import InvoiceNodeName


class RunProgressModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)


class NodeProgress(RunProgressModel):
    name: InvoiceNodeName
    observed_at: AwareDatetime | None = None
    event_type: str | None = None
    state: dict[str, JsonValue] = Field(default_factory=dict)


class RunProgress(RunProgressModel):
    run_id: UUID
    invoice_id: UUID
    status: RunStatus
    graph_version: str
    active_node: InvoiceNodeName | None
    progress_source: Literal["audit-ledger"] = "audit-ledger"
    nodes: list[NodeProgress] = Field(min_length=11, max_length=11)
    started_at: AwareDatetime | None
    completed_at: AwareDatetime | None
    read_at: AwareDatetime
